### PR TITLE
[MRG +1]: construct Montage out of channels with position when reading eeglab files

### LINKED ISSUE #3957: reading eeglab files with channels missing position
I have many eeglab files that contain at least one channel that has undefined position (for example biosemi EXG channels) - the way `mne.io.eeglab._get_info` works currently is to ignore all channel positions if at least one channel has no specified position. 
This behavior originates [here](https://github.com/mne-tools/mne-python/blob/master/mne/io/eeglab/eeglab.py#L81) - `np.unique` of x, y and z position is checked and if the size of the unique is == 1 the channel positions are ignored. In the case of no position for a given channel - `_to_loc` which is used to get channel positions returns 0 for x, y and z - in which case unique is of size 1 and all channel positions are ignored.
I think it would be better if `_to_loc` returned `np.nan` when given value was not present and channels with at least one `np.nan` in position were omitted when building Montage. I can make a PR if you agree.

### PR DESCRIPTION
**Pull request by mmagnuski**
fixes #3957

### CODE REVIEW

**Comment by mmagnuski** [comment]
this probably requires an additional test so I'll add one tomorrow.

**Comment by jasmainak** [comment]
Let me know when you are ready. I have some files locally to test things

**Comment by Eric89GXL** [comment]
I think we usually do `(0., 0., 0.)` to mean "invalid position" (even though technically it could be a real position). `np.nan` is probably also okay though

**Comment by mmagnuski** [comment]
Ah, ok, if you are using 0, 0, 0 as no position then the solution is easier - I'll change it.

**Comment by mmagnuski** [comment]
Hm, I changed to a super-easy solution (locally) when everything is put into montage - all channels, but some with zeros for all x, y and z. But I am getting a weird error when testing locally:
```
ValueError: None of the sensors defined in the montage were found in the info structure. Check the channel names.
```
which wouldn't be strange if not for the fact that Info is:
```
<Info | 15 non-empty fields
    bads : 'list | 0 items
    ch_names : 'list | FP1, Fz
...
```
and Montage:
```
<Montage | user_defined - 2 channels: FP1, Fz ...>
```
so the channels should agree...
Full traceback below:
```
Traceback (most recent call last):
  File "test_eeglab.py", line 156, in <module>
    run_tests_if_main()
  File "c:\proj\src\mne-python\mne\utils.py", line 2199, in run_tests_if_main
    val()
  File "C:\Anaconda3\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "C:\Anaconda3\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "test_eeglab.py", line 85, in test_io_set
    epochs = read_epochs_eeglab(epochs_fname)
  File "c:\proj\src\mne-python\mne\io\eeglab\eeglab.py", line 249, in read_epochs_eeglab
    uint16_codec=uint16_codec)
  File "<string>", line 2, in __init__
  File "c:\proj\src\mne-python\mne\utils.py", line 709, in verbose
    return function(*args, **kwargs)
  File "c:\proj\src\mne-python\mne\io\eeglab\eeglab.py", line 536, in __init__
    info = _get_info(eeg, montage, eog=eog)
  File "c:\proj\src\mne-python\mne\io\eeglab\eeglab.py", line 101, in _get_info
    update_ch_names=update_ch_names)
  File "c:\proj\src\mne-python\mne\io\base.py", line 2340, in _check_update_montage
    _set_montage(info, montage, update_ch_names=update_ch_names)
  File "c:\proj\src\mne-python\mne\channels\montage.py", line 735, in _set_montage
    raise ValueError('None of the sensors defined in the montage were '
ValueError: None of the sensors defined in the montage were found in the info structure. Check the channel names.
```

**Comment by mmagnuski** [comment]
Anyway - the current solution (the one that is present here) passes locally and works for me. The only issue is that if some eeglab chanels do not have defined position it gives a warning (from _apply_montage) that not all channels were found in Montage. For me it is ok - but this may break some tests that count warnings. Is such warning ok for you?

**Comment by mmagnuski** [comment]
@jasmainak @Eric89GXL 
CIs are green, you can take a look.

**Comment by Eric89GXL** [comment]
This seems okay to me. Tests that break should probably be made more robust anyway, e.g. by looking for warnings of interest

File: mne/io/eeglab/tests/test_eeglab.py [hunk 1]
```diff
@@ -135,6 +135,39 @@ def test_io_set():
...
+    one_chanpos_fname = op.join(temp_dir, 'test_chanpos.set')
+    io.savemat(one_chanpos_fname, {'EEG':
+               {'trials': eeg.trials, 'srate': eeg.srate,
+                'nbchan': 3, 'data': np.random.random((3, 3)),
+                'epoch': eeg.epoch, 'event': eeg.epoch,
+                'chanlocs': chanlocs, 'times': eeg.times[:3], 'pnts': 3}})
+    # load it
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter('always')
+        raw = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
```
Comment (jasmainak) [inline mne/io/eeglab/tests/test_eeglab.py]: can you add a test where the user provides a montage and there are missing channels? It should use the montage that the user provides, not create one of its own

**Comment by jasmainak** [comment]
Fix looks reasonable but maybe a test with different settings of montage would be nice. Because it's been a while looking at this code for me. And the code path is not trivially simple for me to verify. At least it doesn't break any of the data files that I tried locally

File: mne/io/eeglab/tests/test_eeglab.py (hunk 1 above)
Comment (mmagnuski) [inline mne/io/eeglab/tests/test_eeglab.py]: sure, I'll add that

**Comment by mmagnuski** [comment]
@jasmainak 
You're right - I didn't consider scenarios with channel positions in set file and montage as a separate file. What should be done if some channels have positions in set but not in montage file? Should such positions be ignored when montage file is passed? That's the easiest solution - so I think I'll add that.

**Comment by mmagnuski** [comment]
@Eric89GXL - how can I abort two previous CIs so that they run only for the last commit? Or if I can't - if you could do that, it would save us a few hours :)

**Comment by jaeilepp** [comment]
>  how can I abort two previous CIs so that they run only for the last commit? Or if I can't - if you could do that, it would save us a few hours :)

Canceled.

**Comment by mmagnuski** [comment]
Thanks @jaeilepp!

File: mne/io/eeglab/eeglab.py [hunk 2]
```diff
@@ -55,34 +55,47 @@ def _to_loc(ll):
...
     # add the ch_names and info['chs'][idx]['loc']
     path = None
     if not isinstance(eeg.chanlocs, np.ndarray) and eeg.nbchan == 1:
             eeg.chanlocs = [eeg.chanlocs]
 
     if len(eeg.chanlocs) > 0:
-        ch_names, pos = list(), list()
+        pos_fields = ['X', 'Y', 'Z']
+        if (isinstance(eeg.chanlocs, np.ndarray) and not
+            isinstance(eeg.chanlocs[0], io.matlab.mio5_params.mat_struct)):
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: sorry one last question. What version of scipy do you need for this check? I hope it's not too recent ...

File: mne/io/eeglab/tests/test_eeglab.py [hunk 3]
```diff
@@ -135,6 +135,72 @@ def test_io_set():
...
+                'chanlocs': nopos_chanlocs, 'times': eeg.times[:2], 'pnts': 2}})
+    # load the file
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter('always')
+        raw = read_raw_eeglab(input_fname=nopos_fname, preload=True)
+    # test that channel names have been loaded but not channel positions
+    for i in range(3):
+        assert_equal(raw.info['chs'][i]['ch_name'], ch_names[i])
+        assert_array_equal(raw.info['chs'][i]['loc'][:3],
+                           np.array([0., 0., 0.,]))
```
Comment (jasmainak) [inline mne/io/eeglab/tests/test_eeglab.py]: this is great. Thanks for all the new tests!

File: mne/io/eeglab/eeglab.py (hunk 2 above)
Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: Frankly - I don't know. But that should be resolved when the tests finish, right? I'll try to check in the meantime.

Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: It seems `mat_struct` has been in scipy already in may of 2009 (although I am not sure if in was present in the same place):
http://scipy-user.10969.n7.nabble.com/io-loadmat-and-mat-struct-object-td11508.html
that would mean that scipy >= 0.7 should have it - so we are safe :)

**Comment by jasmainak** [comment]
okay LGTM +1 for merge when CIs are happy!

**Comment by mmagnuski** [comment]
First Travis job kind of fails (although it runs to completion and all relevant tests pass) due to mayavi problems:
```
Test CombineMarkersModel Traits Model ... Exception occurred in traits notification handler for object: <mayavi.tools.mlab_scene_model.MlabSceneModel object at 0x7ffaf1197650>, trait: do_render, old value: <undefined>, new value: True
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/traits/trait_notifiers.py", line 519, in _dispatch_change_event
    self.dispatch( handler, *args )
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/traits/trait_notifiers.py", line 482, in dispatch
    handler( *args )
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/tvtk/pyface/ui/qt4/scene.py", line 348, in render
    self._vtk_control.Render()
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/tvtk/pyface/ui/qt4/QVTKRenderWindowInteractor.py", line 498, in Render
    self.update()
RuntimeError: wrapped C/C++ object of type _VTKRenderWindowInteractor has been deleted
Exception occurred in traits notification handler for object: <mayavi.tools.mlab_scene_model.MlabSceneModel object at 0x7ffaf1197650>, trait: do_render, old value: <undefined>, new value: True
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/traits/trait_notifiers.py", line 519, in _dispatch_change_event
    self.dispatch( handler, *args )
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/traits/trait_notifiers.py", line 482, in dispatch
    handler( *args )
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/tvtk/pyface/ui/qt4/scene.py", line 348, in render
    self._vtk_control.Render()
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/tvtk/pyface/ui/qt4/QVTKRenderWindowInteractor.py", line 498, in Render
    self.update()
( ... )
```
and some other unrelated problems.

**Comment by Eric89GXL** [comment]
Cool, more `mayavi`-related CI problems :(

**Comment by mmagnuski** [comment]
@Eric89GXL 
Actually it's a bit more fun :) - same/similar errors are present in another job that finishes green. Both perform all tests ok (without failures according to summary at the end) - the only major differences I see is that one is marked as green and other as red... And there is another red (but not failing!) job that does not have these errors. So mayavi is not the reason the builds are red - neither any of the tests... Maybe you can spot the differences - it's a fun game to play ;)

**Comment by mmagnuski** [comment]
But I see some new flake errors that I'll fix - so this is not yet ready.

**Comment by mmagnuski** [comment]
@Eric89GXL @jasmainak 
Ok, it's ready on my end - the builds are marked as failing, but reason for this is not clear to me (as I mentioned in more details above) - all tests pas....

**Comment by agramfort** [comment]
Travis complains

**Comment by mmagnuski** [comment]
@agramfort
See my previous comments - other PRs have the same problem: tests pass but Travis complains.

File: mne/io/eeglab/eeglab.py [hunk 4]
```diff
@@ -55,34 +55,48 @@ def _to_loc(ll):
...
     if len(eeg.chanlocs) > 0:
-        ch_names, pos = list(), list()
+        pos_fields = ['X', 'Y', 'Z']
+        if (isinstance(eeg.chanlocs, np.ndarray) and not isinstance(
+                eeg.chanlocs[0], io.matlab.mio5_params.mat_struct)):
+            has_pos = all([fld in eeg.chanlocs[0].dtype.names
+                           for fld in pos_fields])
+        else:
+            has_pos = all([hasattr(eeg.chanlocs[0], fld)
+                           for fld in pos_fields])
```
Comment (Eric89GXL) [inline mne/io/eeglab/eeglab.py]: does not need `[]` in fact it's better not to have it as it will  short-circuit if possible

**Comment by mmagnuski** [comment]
Travis is sane again and all builds except one came back green ([link](https://travis-ci.org/mne-tools/mne-python/builds/199188183)). The one that is failing was because `URLError: <urlopen error [Errno 101] Network is unreachable>` during test for downloading file over http.

**Comment by jasmainak** [comment]
LGTM, merged. Thanks @mmagnuski !

**Comment by mmagnuski** [comment]
Thanks @jasmainak and @Eric89GXL! :rocket:

**Pull request merged**